File: android/src/main/kotlin/com/example/flutter_pytorch_coreml_cyclegan/channelMethods/CMImageTransform.kt
package com.example.flutter_pytorch_coreml_cyclegan.channelMethods

import com.example.flutter_pytorch_coreml_cyclegan.pytorch.MLProcessor
import com.example.flutter_pytorch_coreml_cyclegan.pytorch.MLProcessorImpl
import com.example.flutter_pytorch_coreml_cyclegan.utils.BitmapProcessor
import com.example.flutter_pytorch_coreml_cyclegan.utils.BitmapProcessorImpl

object CMImageTransform {

    private val bitmapProcessor: BitmapProcessor = BitmapProcessorImpl()
    private var mlProcessor: MLProcessor? = null

    fun setModel(modelPath: String, width: Int, height: Int): String? {
        try {
            mlProcessor = MLProcessorImpl(modelPath, width, height)
        } catch (ex: Exception) {
            return ex.message ?: "unknown"
        }
        return null
    }

    fun transformImage(
        imagePath: String, outputPath: String): String? {
        try {
            mlProcessor?.let {
                val srcBitmap = bitmapProcessor.loadImage(imagePath)
                val input = bitmapProcessor.centerCropScale(
                    srcBitmap, it.width, it.height)
                it.process(input).also {
                    bitmapProcessor.saveBitmap(it, outputPath)
                    return null
                }
            }
            throw Exception("a model is not set")
        } catch (ex: Exception) {
            return ex.message ?: "unknown"
        }
    }
}

File: ios/Classes/ChannelMethods/CMImageTransform.swift
//
//  CMImageTransform.swift
//  flutter_ml_image_transformation
//
//  Created by no145 on 2022/01/08.
//

import Foundation
import UIKit

class CMImageTransform {
    
    static let imageUtils: ImageUtils = ImageUtilsImpl()
    static var mlProcessor: MLProcessor?
    
    static func setModel(modelPath: String, width: Int, height: Int) -> String? {
        do {
            if (mlProcessor == nil) {
                mlProcessor = try MLProcessorImpl(modelPath, width, height)
            }
            return nil
        } catch let error {
            return error.localizedDescription
        }
    }
    
    static func transformImage(
        imagePath: String, outputPath: String) -> String? {
        do {
            guard let mlProcessor = self.mlProcessor else {
                throw AppError.nilResult("a model is not set")
            }
            let src = try imageUtils.loadImage(path: imagePath)
            let input = src.centerCropScale(
                width: mlProcessor.width, height: mlProcessor.height)
            let output = try mlProcessor.process(image: input)
            try imageUtils.saveImage(image: output, path: outputPath)
            return nil
        } catch let error {
            return error.localizedDescription
        }
    }
}
